File: backend/app/schemas/documents.py
import uuid
from datetime import datetime
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.enums import DocumentStatus, DocumentType


class DocumentSortBy(str, Enum):
    ORIGINAL_FILE_NAME = "original_file_name"
    DOCUMENT_TYPE = "document_type"
    STATUS = "status"
    UPLOADED_AT = "uploaded_at"
    FILE_SIZE = "file_size"


class SortOrder(str, Enum):
    ASC = "asc"
    DESC = "desc"


class DocumentUploadForm(BaseModel):
    client_company_id: uuid.UUID
    document_type: DocumentType
    document_category: str | None = Field(default=None, max_length=100)
    note: str | None = None

    @field_validator("document_category", "note", mode="before")
    @classmethod
    def normalize_optional_text(cls, raw_value: str | None) -> str | None:
        if raw_value is None:
            return None

        normalized_value = raw_value.strip()
        return normalized_value or None


class DocumentClientCompanySummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company_name: str
    tax_code: str | None
    is_active: bool


class DocumentUploaderSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    email: str
    full_name: str


class DocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_company_id: uuid.UUID
    uploaded_by_user_id: uuid.UUID
    original_file_name: str
    stored_file_name: str
    file_size: int
    mime_type: str
    document_type: DocumentType
    document_category: str | None
    status: DocumentStatus
    note: str | None
    uploaded_at: datetime
    created_at: datetime
    updated_at: datetime


class DocumentListItem(DocumentRead):
    client_company: DocumentClientCompanySummary
    uploaded_by: DocumentUploaderSummary


class DocumentDetail(DocumentListItem):
    pass


class DocumentListResponse(BaseModel):
    records: list[DocumentListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


class OcrResultRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    raw_text: str | None
    confidence_score: Decimal | None
    processed_at: datetime | None
    created_at: datetime
    updated_at: datetime

File: frontend/types/documents.ts
import type { AuthenticatedUserApiResponse, UserRole } from "@/types/auth";

export type DocumentType =
  | "VAT_INVOICE"
  | "RECEIPT"
  | "PAYMENT_VOUCHER"
  | "IMPORT_WAREHOUSE"
  | "EXPORT_WAREHOUSE"
  | "OTHER";

export type DocumentStatus =
  | "UPLOADED"
  | "PROCESSING"
  | "OCR_DONE"
  | "REVIEWED"
  | "APPROVED"
  | "FAILED";

export type DocumentClientCompanySummary = {
  id: string;
  companyName: string;
  taxCode: string | null;
  isActive: boolean;
};

export type DocumentUploaderSummary = {
  id: string;
  email: string;
  fullName: string;
};

export type DocumentClientCompanySummaryApiResponse = {
  id: string;
  company_name: string;
  tax_code: string | null;
  is_active: boolean;
};

export type DocumentUploaderSummaryApiResponse = Pick<
  AuthenticatedUserApiResponse,
  "id" | "email" | "full_name"
>;

export type DocumentApiResponse = {
  id: string;
  client_company_id: string;
  uploaded_by_user_id: string;
  original_file_name: string;
  stored_file_name: string;
  file_size: number;
  mime_type: string;
  document_type: DocumentType;
  document_category: string | null;
  status: DocumentStatus;
  note: string | null;
  uploaded_at: string;
  created_at: string;
  updated_at: string;
};

export type DocumentListItemApiResponse = DocumentApiResponse & {
  client_company: DocumentClientCompanySummaryApiResponse;
  uploaded_by: DocumentUploaderSummaryApiResponse;
};

export type DocumentDetailApiResponse = DocumentListItemApiResponse;

export type UploadedDocument = {
  id: string;
  clientCompanyId: string;
  uploadedByUserId: string;
  originalFileName: string;
  storedFileName: string;
  fileSize: number;
  mimeType: string;
  documentType: DocumentType;
  documentCategory: string | null;
  status: DocumentStatus;
  note: string | null;
  uploadedAt: string;
  createdAt: string;
  updatedAt: string;
};

export type DocumentListItem = UploadedDocument & {
  clientCompany: DocumentClientCompanySummary;
  uploadedBy: DocumentUploaderSummary;
};

export type DocumentDetail = DocumentListItem;

export type DocumentListApiResponse = {
  records: DocumentListItemApiResponse[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
};

export type DocumentListResponse = {
  records: DocumentListItem[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
};

export type DocumentListQuery = {
  search?: string;
  clientCompanyId?: string;
  documentType?: DocumentType;
  status?: DocumentStatus;
  uploadedFrom?: string;
  uploadedTo?: string;
  page: number;
  pageSize: number;
  sortBy: "original_file_name" | "document_type" | "status" | "uploaded_at" | "file_size";
  sortOrder: "asc" | "desc";
};

export type DocumentUploadPayload = {
  file: File;
  clientCompanyId: string;
  documentType: DocumentType;
  documentCategory: string;
  note: string;
};

export type DocumentDownload = {
  blob: Blob;
  fileName: string;
  mimeType: string;
};

export type DocumentActionContext = {
  currentUserId: string;
  currentUserRole: UserRole;
};
